File: Management/Services/WebRequestor.cs
using System.Net;
using Microsoft.Extensions.Configuration;
using RestSharp;

public class WebRequestor : IWebRequestor
{
  private string BaseUrl = "";
  private string token;
  private RestClient client;
  private readonly IConfiguration _config;
  private readonly ILogger<WebRequestor> logger;
  private static int rateLimitRetryCount = 6;
  private static int rateLimitSleepInterval = 1000;

  public WebRequestor(IConfiguration config, ILogger<WebRequestor> logger)
  {
    _config = config;
    this.logger = logger;
    token =
      _config["CANVAS_TOKEN"]
      ?? throw new Exception("CANVAS_TOKEN not in environment");
    BaseUrl = _config["CANVAS_URL"] + "/api/v1/";
    client = new RestClient(BaseUrl);
    client.AddDefaultHeader("Authorization", $"Bearer {token}");
  }

  public async Task<(T[]?, RestResponse)> GetManyAsync<T>(RestRequest request)
  {
    var response = await client.ExecuteGetAsync(request);
    return (deserialize<T[]>(response), response);
  }

  public async Task<(T?, RestResponse)> GetAsync<T>(RestRequest request)
  {
    var response = await client.ExecuteGetAsync(request);
    return (deserialize<T>(response), response);
  }

  public async Task<RestResponse> PostAsync(RestRequest request) => await rateLimitAwarePostAsync(request, 0);


  private async Task<RestResponse> rateLimitAwarePostAsync(RestRequest request, int retryCount = 0)
  {
    request.AddHeader("Content-Type", "application/json");

    var response = await client.ExecutePostAsync(request);

    if (isRateLimited(response))
    {
      if (retryCount < rateLimitRetryCount)
      {
        logger.LogInformation($"hit rate limit on post, retry count is {retryCount} / {rateLimitRetryCount}, retrying");
        Console.WriteLine($"hit rate limit on post, retry count is {retryCount} / {rateLimitRetryCount}, retrying");
        Thread.Sleep(rateLimitSleepInterval);
        return await rateLimitAwarePostAsync(request, retryCount + 1);
      }
    }

    if (!response.IsSuccessful)
    {
      logger.LogError($"Error with response, response content: {response.Content}");
      throw new Exception($"error post response, retrycount: {retryCount}, ratelimited: {isRateLimited(response)}, code: {response.StatusCode}, response content: {response.Content}");
    }
    return response;
  }

  private static bool isRateLimited(RestResponse response)
  {
    if (response.Content == null)
      return false;
    return response.StatusCode == HttpStatusCode.Forbidden
      && response.Content.Contains("403 Forbidden (Rate Limit Exceeded)");
  }

  public async Task<(T?, RestResponse)> PostAsync<T>(RestRequest request)
  {
    var response = await PostAsync(request);
    return (deserialize<T>(response), response);
  }

  public async Task<RestResponse> PutAsync(RestRequest request)
  {
    request.AddHeader("Content-Type", "application/json");
    var response = await client.ExecutePutAsync(request);
    // if (!response.IsSuccessful)
    // {
    //   Console.WriteLine(response.Content);
    //   Console.WriteLine(response.ResponseUri);
    //   Console.WriteLine("error with response");
    //   throw new Exception("error with response");
    // }
    return response;
  }

  public async Task<(T?, RestResponse)> PutAsync<T>(RestRequest request)
  {
    request.AddHeader("Content-Type", "application/json");
    var response = await client.ExecutePutAsync(request);
    return (deserialize<T>(response), response);
  }

  public Task<RestResponse> DeleteAsync(RestRequest request) => recursiveDeleteAsync(request, 0);
  private async Task<RestResponse> recursiveDeleteAsync(RestRequest request, int retryCount)
  {
    try
    {
      var response = await client.DeleteAsync(request);
      if (isRateLimited(response))
        Console.WriteLine("after delete response in rate limited");
      return response;
    }
    catch (HttpRequestException e)
    {
      if (e.StatusCode == HttpStatusCode.Forbidden) // && response.Content == "403 Forbidden (Rate Limit Exceeded)"
      {
        if (retryCount < rateLimitRetryCount)
        {
          logger.LogInformation($"hit rate limit in delete, retry count is {retryCount} / {rateLimitRetryCount}, retrying");
          Console.WriteLine($"hit rate limit in delete, retry count is {retryCount} / {rateLimitRetryCount}, retrying");
          Thread.Sleep(rateLimitSleepInterval);
          return await recursiveDeleteAsync(request, retryCount + 1);
        }
        else
        {
          logger.LogInformation($"hit rate limit in delete, {rateLimitRetryCount} retries did not fix it");
        }
      }
      throw;
    }
  }

  private static T? deserialize<T>(RestResponse response)
  {
    // using var activity = DiagnosticsConfig.Source.StartActivity("deserializing response");
    // activity?.AddTag("url", response.ResponseUri);

    if (!response.IsSuccessful)
    {
      Console.WriteLine(response.Content);
      Console.WriteLine(response.ResponseUri);
      Console.WriteLine(response.ErrorMessage);
      Console.WriteLine("error with response");
      // Console.WriteLine(JsonSerializer.Serialize(response));
      Console.WriteLine(JsonSerializer.Serialize(response.Request?.Parameters));
      throw new Exception($"error with response to {response.ResponseUri} {response.StatusCode}");
    }
    try
    {
      var data = JsonSerializer.Deserialize<T>(response.Content!);

      if (data == null)
      {
        Console.WriteLine(response.Content);
        Console.WriteLine(response.ResponseUri);
        Console.WriteLine("could not parse response, got empty object");
      }
      return data;
    }
    catch (NotSupportedException)
    {
      Console.WriteLine(response.Content);
      throw;
    }
    catch (JsonException)
    {
      Console.WriteLine(response.ResponseUri);
      Console.WriteLine(response.Content);
      Console.WriteLine($"An error occurred during deserialization");
      throw;
    }
  }

}

File: nextjs/src/services/canvas/webRequestor.ts
type FetchOptions = Omit<RequestInit, "method">;

const token = process.env.CANVAS_TOKEN;
if (!token) {
  throw new Error("CANVAS_TOKEN not in environment");
}

const baseUrl = `${process.env.CANVAS_URL}/api/v1/`;
const rateLimitRetryCount = 6;
const rateLimitSleepInterval = 1000;

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

const isRateLimited = async (response: Response): Promise<boolean> => {
  const content = await response.text();
  return (
    response.status === 403 &&
    content.includes("403 Forbidden (Rate Limit Exceeded)")
  );
};

const deserialize = async <T>(response: Response): Promise<T | undefined> => {
  if (!response.ok) {
    console.error(`Error with response to ${response.url} ${response.status}`);
    throw new Error(
      `Error with response to ${response.url} ${response.status}`
    );
  }
  try {
    return (await response.json()) as T;
  } catch (e) {
    console.error(`An error occurred during deserialization: ${e}`);
    throw e;
  }
};

const rateLimitAwarePost = async (
  url: string,
  body: any,
  retryCount = 0
): Promise<Response> => {
  const response = await fetch(url, {
    method: "POST",
    body: JSON.stringify(body),
    headers: {
      Authorization: `Bearer ${token}`,
      "Content-Type": "application/json",
    },
  });

  if (await isRateLimited(response)) {
    if (retryCount < rateLimitRetryCount) {
      console.info(
        `Hit rate limit on post, retry count is ${retryCount} / ${rateLimitRetryCount}, retrying`
      );
      await sleep(rateLimitSleepInterval);
      return await rateLimitAwarePost(url, body, retryCount + 1);
    }
  }

  if (!response.ok) {
    const content = await response.text();
    console.error(`Error with response, response content: ${content}`);
    throw new Error(
      `Error post response, retrycount: ${retryCount}, ratelimited: ${await isRateLimited(
        response
      )}, code: ${response.status}, response content: ${content}`
    );
  }
  return response;
};

const recursiveDelete = async (
  url: string,
  options: FetchOptions,
  retryCount = 0
): Promise<Response> => {
  try {
    const response = await fetch(url, {
      ...options,
      method: "DELETE",
      headers: {
        ...options.headers,
        Authorization: `Bearer ${token}`,
      },
    });

    if (await isRateLimited(response)) {
      console.info("After delete response in rate limited");
      await sleep(rateLimitSleepInterval);
      return await recursiveDelete(url, options, retryCount + 1);
    }

    return response;
  } catch (e) {
    const error = e as Error & { response?: Response };
    if (error.response?.status === 403) {
      if (retryCount < rateLimitRetryCount) {
        console.info(
          `Hit rate limit in delete, retry count is ${retryCount} / ${rateLimitRetryCount}, retrying`
        );
        await sleep(rateLimitSleepInterval);
        return await recursiveDelete(url, options, retryCount + 1);
      } else {
        console.info(
          `Hit rate limit in delete, ${rateLimitRetryCount} retries did not fix it`
        );
      }
    }
    throw e;
  }
};
export const webRequestor = {
  getMany: async <T>(url: string, options: FetchOptions = {}) => {
    const response = await fetch(url, {
      ...options,
      method: "GET",
      headers: {
        ...options.headers,
        Authorization: `Bearer ${token}`,
      },
    });
    return { data: await deserialize<T[]>(response), response };
  },

  get: async <T>(url: string, options: FetchOptions = {}) => {
    const response = await fetch(url, {
      ...options,
      method: "GET",
      headers: {
        ...options.headers,
        Authorization: `Bearer ${token}`,
      },
    });
    return { data: await deserialize<T>(response), response };
  },

  post: async (url: string, body: any) => {
    return await rateLimitAwarePost(url, body);
  },

  postWithDeserialize: async <T>(url: string, body: any) => {
    const response = await rateLimitAwarePost(url, body);
    return { data: await deserialize<T[]>(response), response };
  },

  put: async (url: string, body: any = {}) => {
    const response = await fetch(url, {
      method: "PUT",
      body: JSON.stringify(body),
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
    return response;
  },

  putWithDeserialize: async <T>(url: string, body: any = {}) => {
    const response = await fetch(url, {
      body: JSON.stringify(body),
      method: "PUT",
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
    return { data: await deserialize<T[]>(response), response };
  },

  delete: async (
    url: string,
    options: FetchOptions = {}
  ): Promise<Response> => {
    return await recursiveDelete(url, options);
  },
};
